pkg/nsx/services/securitypolicy/firewall.go

```diff
@@ -927,43 +872,135 @@ func (service *SecurityPolicyService) getUpdateShares(existingShares []*model.Sh
+		deleteShares = append(deleteShares, *share)
 	}
-	for i := len(*deleteShares) - 1; i >= 0; i-- {
-		(*deleteShares)[i].MarkedForDelete = &MarkedForDelete
+	for i := len(deleteShares) - 1; i >= 0; i-- {
+		(deleteShares)[i].MarkedForDelete = &MarkedForDelete
+	}
+	return deleteShares
+}
+
+func (service *SecurityPolicyService) getStaleUpdateShares(nsxShares []model.Share) (staleShares []model.Share, updatedShares []model.Share) {
+	finalStaleShares := make([]model.Share, 0)
+	finalChangedShares := make([]model.Share, 0)
+	for i := len(nsxShares) - 1; i >= 0; i-- {
+		if nsxShares[i].MarkedForDelete != nil && (*nsxShares[i].MarkedForDelete == MarkedForDelete) {
+			finalStaleShares = append(finalStaleShares, nsxShares[i])
+		} else {
+			finalChangedShares = append(finalChangedShares, nsxShares[i])
+		}
+	}
+	return finalStaleShares, finalChangedShares
+}
+
+func (service *SecurityPolicyService) getStaleUpdateGroups(nsxGroups []model.Group) (staleShares []model.Group, updatedShares []model.Group) {
+	finalStaleGroups := make([]model.Group, 0)
+	finalChangedGroups := make([]model.Group, 0)
+
+	for i := len(nsxGroups) - 1; i >= 0; i-- {
+		if nsxGroups[i].MarkedForDelete != nil && (*nsxGroups[i].MarkedForDelete == MarkedForDelete) {
+			finalStaleGroups = append(finalStaleGroups, nsxGroups[i])
+		} else {
+			finalChangedGroups = append(finalChangedGroups, nsxGroups[i])
+		}
+	}
+	return finalStaleGroups, finalChangedGroups
+}
+
+func (service *SecurityPolicyService) markSecurityPolicyResourcesDelete(indexScope string, spUID types.UID) (
+	*model.SecurityPolicy, []model.Group, []model.Share, []model.Group, []model.Share, []model.Group, common.VPCResourceInfo, error,
+) {
+	var nsxSecurityPolicy *model.SecurityPolicy
+	var err error
+
+	securityPolicyStore, ruleStore, groupStore := service.getSecurityPolicyResourceStores()
+	infraGroupStore, infraShareStore, projectGroupStore, projectShareStore := service.getVPCShareResourceStores()
+
+	existingSecurityPolices := securityPolicyStore.GetByIndex(indexScope, string(spUID))
+	// There is no NSX groups/rules in the security policy retrieved from securityPolicy store.
+	// The groups/rules associated the deleting security policy can only be gotten from group/rule store.
+	existingGroups := groupStore.GetByIndex(indexScope, string(spUID))
+	existingRules := ruleStore.GetByIndex(indexScope, string(spUID))
+	nsxGroups := service.getMarkDeleteGroups(existingGroups, spUID)
+	nsxRules := service.getMarkDeleteRules(existingRules, spUID)
+
+	// For other GC cases, it will follow the normal deletion process as below.
+	var vpcInfo common.VPCResourceInfo
+	if len(existingSecurityPolices) != 0 {
+		nsxSecurityPolicy = existingSecurityPolices[0]
+		// vpcInfo should be listed directly from security policy store to avoid calling VPC service.
+		// Get orgID, projectID, vpcID from security policy path "/orgs/<orgID>/projects/<projectID>/vpcs/<vpcID>/security-policies/<spID>"
+		if nsxSecurityPolicy.Path == nil {
```

Comment:
In which case the `nsxSecurityPolicy.Path` is `nil`? Or it's only a preventive check?

Reply:
It's more for a preventive check, do you think we can remove this check.